### DISCUSSION TITLE
Speech Recognition | Seeed Studio Wiki

### DISCUSSION THREAD

**Discussion by giscus** [Seeed-Studio/wiki-documents#192] [Q&A] [answered]
# Speech Recognition | Seeed Studio Wiki

Speech Recognition

https://wiki.seeedstudio.com/XIAO-BLE-Sense-TFLite-Mic/

**Comment by stanleyoz** [comment]
Nice write up! We getting this error after uploading our file over to the Sense board ..

Feature generation failed
Requested feature_data_ size 536871936 doesn't match 1960

Anyone can point to where I probably messed up along the way? Thanks

**Comment by MatthewJeffson** [comment]
Welcome to our product documentation platform! We value user feedback and contributions, so we couldn’t be more grateful that you can help us improving the documentation.

Hence, if you find obvious errors in the product documentation, or you have additional ideas and insights about using products, please [contribute](https://wiki.seeedstudio.com/Contribution-Guide) to us.

If you need technical support, you can email it to [fae@seeed.cc](mailto:fae@seeed.cc).

If you want to discuss the product, I recommend the [Seeed Studio Forum](https://forum.seeedstudio.com/) and [Seeed Studio Discord Channel](https://discord.gg/eWkprNDMU7)

The "Discussion channel" in Github here is a relatively new addition and may not receive much attention.

Nonetheless, we hope you find it helpful with the help of others. We will clean it up on a monthly basis.

**Comment by ahsanfi** [comment]
Hello, I'm trying to record audio (.wav) to sd card with XIAO BLE SENSE nRF5840 with seeduino extension, but keep getting failed like this:

Capturing .wav samples
initialization failed!

I think the module cannot detect the sdcard. Any solutions? Thanks

**Reply by xzxcessarr** [reply]
Hello, Could you tell us how much space of your tf card and test the PDM microphone by the example mic_serial_plotter? This error may caused by the tf card or the PDM microphone, we need futher locate the error by your help

**Comment by ahsanfi** [comment, marked as answer]
My problem has solved. I think it because the SD Card capacity. I get error when I use 64GB sd card, then I change it to 16 GB. And its work

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by xzxcessarr** [reply]
Thanks for feedback!

**Comment by javiBajoCero** [comment]
If someone else is getting 
"error: conflicting declaration 'unsigned int g_model_len'
 unsigned int g_model_len = 19160;"

Is because we blindly copypasted everything including g_model_len at the end of model.cc

The python script is declaring g_model_len as unsigned.

You fix this by removing the unsigned in unsigned int g_model_len = 19160;" inside the cpp file

**Reply by javiBajoCero** [reply]
Alright that will make it compile but this error will show in the serial monitor

"Feature generation failed" 
"Requested feaure_data_size 536871936 doesnt match 1960"


That feels like a signed integer overflow? or i messed something up with the copypasting

**Reply by javiBajoCero** [reply]
This guy had the same problem and fixed it , https://github.com/tensorflow/tensorflow/issues/61074

**Reply by javiBajoCero** [reply]
Also this guy https://github.com/tensorflow/tensorflow/issues/44868

**Reply by nfs0619** [reply]
Thank you for your valuable feedback and for sharing the solution with us! We truly appreciate you taking the time to help improve the experience for everyone. If you encounter any other issues or have suggestions, please don’t hesitate to reach out.

**Comment by valentas** [comment]
This is not working on PlatformIO

tried to compile example but stuck with some (PDMFilter_CM7_GCC) lib not found.

log at the end:
Archiving .pio/build/stable/libFrameworkArduino.a
Indexing .pio/build/stable/libFrameworkArduino.a
Archiving .pio/build/stable/libed2/libtflite-micro-arduino-examples.a
Indexing .pio/build/stable/libed2/libtflite-micro-arduino-examples.a
Linking .pio/build/stable/firmware.elf
/home/v/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: cannot find -lPDMFilter_CM7_GCC_wc32
collect2: error: ld returned 1 exit status
*** [.pio/build/stable/firmware.elf] Error 1


platformio.ini:

[env:stable]
platform = nordicnrf52
board = xiaoblesense
framework = arduino
monitor_speed = 115200
build_flags = 
	-DCORE_DEBUG_LEVEL=3

**Reply by valentas** [reply]
I also tried suggested platformio.ini config and got same results:

[env]
platform = https://github.com/maxgerhardt/platform-nordicnrf52
framework = arduino
monitor_speed = 115200
build_flags = 
	-DCORE_DEBUG_LEVEL=3

[env:xiaoblesense_arduinocore_mbed]
board = xiaoblesense
 
[env:xiaoble_arduinocore_mbed]
board = xiaoble

**Reply by nfs0619** [reply]
Hello, the compilation error is due to the missing `PDMFilter_CM7_GCC` library. It seems that the `platformio.ini` configuration may not correctly reference the customized Nordic nRF52 platform. We kindly suggest checking if the board configuration is correct and ensuring that the missing library files (such as `hal_input_pdm_mic.c` and `PDMFilter_CM7_GCC`) are placed in the project's `drivers` or `libs` directory. If the issue persists, please feel free to provide more detailed steps and logs to our technical support email at [techsupport@seeed.io](mailto:techsupport@seeed.io), and they will be happy to assist you further.

**Comment by jasg1612** [comment]
After crying so much and trying with that mistake <<Feature generation failed
Requested feature_data_ size 536871936 doesn't match 1960>>, God brought me to a solution. go to the ino file, which according to the library installed is called <<micro_speech.ino>>
step 01
In this line of code the default code is 4 change to 5
 static tflite::MicroMutableOpResolver<5> micro_op_resolver(error_reporter);
step 02, in these lines of code 
  if (micro_op_resolver.AddDepthwiseConv2D() != kTfLiteOk) {
    return;
  }
  if (micro_op_resolver.AddFullyConnected() != kTfLiteOk) {
    return;
  }
  if (micro_op_resolver.AddSoftmax() != kTfLiteOk) {
    return;
  }
  if (micro_op_resolver.AddReshape() != kTfLiteOk) {
    return;
  }

add the code at the end like this
  static tflite::MicroMutableOpResolver<5> micro_op_resolver(error_reporter);
  if (micro_op_resolver.AddDepthwiseConv2D() != kTfLiteOk) {
    return;
  }
  if (micro_op_resolver.AddFullyConnected() != kTfLiteOk) {
    return;
  }
  if (micro_op_resolver.AddSoftmax() != kTfLiteOk) {
    return;
  }
  if (micro_op_resolver.AddReshape() != kTfLiteOk) {
    return;
  }
  if (micro_op_resolver.AddConv2D() != kTfLiteOk) { 
    return;
  } // I have added the Conv2D op and deleted the DepthwiseConv2D op.

**Reply by nfs0619** [reply]
Thanks for sharing!
